experimental/dds/tree/src/id-compressor/IdCompressor.ts

```diff
@@ -1649,17 +1640,15 @@ export class IdCompressor {
 		let attributionId: AttributionId | undefined;
 		let serializedLocalState: SerializedLocalState | undefined;
 		if (newSessionIdMaybe === undefined) {
-			// TODO: This cast can be removed on typescript 4.6
-			const serializedSessionData =
-				serializedSessions[(serialized as SerializedIdCompressorWithOngoingSession).localSessionIndex];
+			const [serializedWithSession] = args;
```

Comment:
Is this destructuring 'args' a second time to avoid the cast?

I'm not sure the readability or runtime impact is worth it, but if kept it probably deserves a similar comment as the 'tmp' assignments.

Reply:
I have added such a comment. I'll consider this good enough for now and if desired @noencke can make a follow up change.

Perf cost here is one array index that should be in the data cache already, and this branch of the if already does a few other indexes, so I expect it to not make a real perf difference.

Reply:
@DLehenbauer The destructuring pattern is (sometimes) useful for the cases where you have a couple overloads of a function, and you can infer which overload is being called by checking the type of one of the arguments. If you actually express the args themselves as a discriminated union, then typescript knows (sometimes) after the check what the _other_ arguments are - without this trick, you have to manually assert or cast them. So above, ts can infer the type of `serializedWithSession` based on what it knows about  `newSessionIdMaybe`. 

Is this a good pattern? Sometimes, IMO. I think with certain function signatures it's a clear win; I'm not convinced that it is here. This instance was kind of an experiment to see how I felt about it over time, and I'm still not sure if I prefer it or not to just doing the casts and being careful. It's technically safer, but the parameters are real ugly.